tests/unit/test_protobuf.py

```diff
@@ -0,0 +1,90 @@
+import pytest
+
+from guidance import (
+    byte_range,
+    char_range,
+    commit_point,
+    select,
+    string,
+    token_limit,
+    with_temperature,
+)
+from guidance._grammar import (
+    Byte,
+    ByteRange,
+    GrammarFunction,
+    Join,
+    ModelVariable,
+    Select,
+)
+
+
+def compare_grammars(g1: GrammarFunction, g2: GrammarFunction) -> bool:
```

Comment:
Feels more or less worthy of being the `__eq__` on grammars!

Rather than hard-coding the fields of each of these classes, could we reach into their `__slots__` in order to make this a bit more robust to changes their implementations? Without actually implementing this as `__eq__`, you'll still have to special-case `values` though.

Just a thought :)

Reply:
I like the idea of using slots instead of hard coding the names of the properties. If we move the code for `compare_grammars` to implement `__eq__`, would we be implementing `__eq__` for the `GrammarFunction` base class, or somewhere else?

Reply:
I think it could go as low level as the `Function` base class, but I'm actually a bit hesitant to implement `__eq__`. Because there is a contract that `a == b` implies `hash(a) == hash(b)`, providing a non-default impl of eq makes objects non-hashable until you provide an impl of that too. I'm fairly certain we use their hashability, so this could potentially break a lot of things if we're not careful